### ISSUE TITLE
Conversion for ClientMessages with format_ != 32 will always fail

### ISSUE THREAD

**Issue by psychon** [sminez/penrose#143]
This code here tries to convert a received client message to penrose's `ClientMessage` type:
https://github.com/sminez/penrose/blob/9f34f030b3435dc399267dda784a969229958b1e/src/xcb/api.rs#L498-L510
`try_from_data()` expects a slice of length 5. This only works for `format_ = 32`. Other formats contain the same number of bytes and thus have more entries. Any such client message will fail conversion (and that was already the case before my recent PR).

Slightly offtopic:
> "ClientMessageEvent.format should really be an enum..." 

That doesn't really solve the problem. Some piece of code somewhere has to deal with the fact that there is a byte that must have a value 8, 16, or 32. No X11 server should send anything else here, but in theory it is possible to receive an invalid value. :-(

**Comment by sminez** [comment]
The enum comment was more that you have to switch on `e.format()` which is a `u8`: if the X server should only ever send data that is `u8`, `u16`, or `u32` aligned then an enum like this makes sense:

```rust
pub enum ClientMessageFormat {
    U8,
    U16,
    U32,
    Invalid,
}
```

as opposed to switching on a raw `u8` and having to specify a `_` catch all :shrug:

(I'd also hope that XCB itself would catch that the format was invalid when attempting to build out the underlying struct...)

As for this failing for slices with len != 5 before the change to make `ClientMessage.data` explicitly `[u32; 5]`, I'm not sure I follow. This should be taking whichever of the three valid formats we received, cast each of the values to `u32` (which should always be safe as we will never be truncating / overflowing) and then storing that as a `Vec` that is the same length as the data received...no? I feel like I'm missing something here...

Also, if this is an issue then should `ClientMessage.data` not be modified as well / instead? (If what was there before isn't correct I'd appreciate a nudge in the right direction as I thought I had it :sweat_smile: )

**Comment by psychon** [comment]
> I feel like I'm missing something here...

Sorry for not being clear. I think the missing piece of information is: A ClientMessage either contains
- `[u32; 5]`, or
- `[u16; 10]`, or
- `[u8; 20]`.

The smaller sizes have more entries than the `u32` message.

The existing code converts these to a `Vec<u32>` that has length 5, 10, or 20. But `ClientMessage::try_from_data` then rejects any `Vec` with a length != 5.

The above is what I should have written instead of the confusing mess I wrote above. Sorry!

**Comment by sminez** [comment]
Aaaaaaah, ok that makes sense now :smile:

**Comment by sminez** [comment]
Fixed in #156
